File: factory_method/factory_method_test.go
package factory__test

import (
	"design_pattern/factory_method/creator"
	"design_pattern/factory_method/product"
	"design_pattern/factory_method/product/part"
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestFactoryMethod(t *testing.T) {
	t.Run("Test Factory Method", func(t *testing.T) {
		superVehicleFactory := creator.NewSuperVehicleFactory()
		ultraVehicleFactory := creator.NewUltraVehicleFactory()

		superBus, err := superVehicleFactory.OrderVehicle(product.Bus)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, product.SuperBusName, superBus.GetName())
		assert.Equal(t, part.EngineHybrid, superBus.GetEngine())
		assert.Equal(t, part.TireNormal, superBus.GetTire())

		superTruck, err := superVehicleFactory.OrderVehicle(product.Truck)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, product.SuperTruckName, superTruck.GetName())
		assert.Equal(t, part.EngineGasoline, superTruck.GetEngine())
		assert.Equal(t, part.TireOffload, superTruck.GetTire())

		ultraBus, err := ultraVehicleFactory.OrderVehicle(product.Bus)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, product.UltraBusName, ultraBus.GetName())
		assert.Equal(t, part.EngineGasoline, ultraBus.GetEngine())
		assert.Equal(t, part.TireNormal, ultraBus.GetTire())

		ultraTruck, err := ultraVehicleFactory.OrderVehicle(product.Truck)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, product.UltraTruckName, ultraTruck.GetName())
		assert.Equal(t, part.EngineHybrid, ultraTruck.GetEngine())
		assert.Equal(t, part.TireOffload, ultraTruck.GetTire())
	})
}

File: typescript/factory-method/factory-method.test.ts
import { SuperVehicleFactory } from "./creator/super-vehicle-factory";
import { UltraVehicleFactory } from "./creator/ultra-vehicle-factory";
import { GASOLINE_ENGINE, HYBRID_ENGINE } from "./product/part/engine";
import { NORMAL_TIRE, OFFLOAD_TIRE } from "./product/part/tire";
import { SuperBus } from "./product/super-bus";
import { SuperTruck } from "./product/super-truck";
import { UltraBus } from "./product/ultra-bus";
import { UltraTruck } from "./product/ultra-truck";
import { BUS, TRUCK } from "./product/vehicle-type";

test('Test Factory Method', () => {
    const superVehicleFactory = new SuperVehicleFactory();
    const ultraVehicleFactory = new UltraVehicleFactory();

    const superBus = superVehicleFactory.orderVehicle(BUS);
    expect(superBus.getName()).toEqual(SuperBus.NAME);
    expect(superBus.getEngine()).toEqual(HYBRID_ENGINE);
    expect(superBus.getTire()).toEqual(NORMAL_TIRE);

    const superTruck = superVehicleFactory.orderVehicle(TRUCK);
    expect(superTruck.getName()).toEqual(SuperTruck.NAME);
    expect(superTruck.getEngine()).toEqual(GASOLINE_ENGINE);
    expect(superTruck.getTire()).toEqual(OFFLOAD_TIRE);

    const ultraBus = ultraVehicleFactory.orderVehicle(BUS);
    expect(ultraBus.getName()).toEqual(UltraBus.NAME);
    expect(ultraBus.getEngine()).toEqual(GASOLINE_ENGINE);
    expect(ultraBus.getTire()).toEqual(NORMAL_TIRE);

    const ultraTruck = ultraVehicleFactory.orderVehicle(TRUCK);
    expect(ultraTruck.getName()).toEqual(UltraTruck.NAME);
    expect(ultraTruck.getEngine()).toEqual(HYBRID_ENGINE);
    expect(ultraTruck.getTire()).toEqual(OFFLOAD_TIRE);
});